src/text_file_functions.cpp

```diff
@@ -722,89 +722,69 @@ int AOApplication::get_saved_ooc_format()
   return result;
 }
 
-QString Courtroom::save_ic_chatlog()
+int Courtroom::save_chatlog(bool is_ooc)
 {
+    QString log_type;
+    if (!is_ooc)
+      log_type = "IC";
+    else
+      log_type = "OOC";
     QString dtstamp_clean = QDateTime::currentDateTimeUtc().toString().replace(" ","-").replace(":","."); // colons are illegal in Windows filenames, and spaces are for chumps
-    QString logfile = "ICLog-" + dtstamp_clean +  ".log";
-    ic_result = ERR_UNKNOWN;
+    QString logfile = "Log" + log_type + '-' + dtstamp_clean +  ".log";
+    save_log_result result = ERR_UNKNOWN;
+
     QDir logfolder("logs");
     if (!logfolder.exists())
     {
       qDebug() << "I: directory ./logs does not exist, attempting to create";
       QDir::current().mkdir(logfolder.dirName());
       if (!logfolder.exists()) { // check again to be sure the folder was created
         qDebug() << "W: attempted to create dir ./logs unsuccessfully";
-        ic_result = ERR_CREATEDIR;
-        return logfile;
+        result = ERR_CREATEDIR;
+        return result;
       }
     }
+
     QFile file(logfolder.dirName() + '/' + logfile);
     if (!file.open(QIODevice::WriteOnly | QIODevice::Text | QIODevice::Truncate))
     {
       qDebug() << "E: could not create log file (" << file.errorString() << ")";
-      ic_result = ERR_CREATELOG;
-      return logfile;
+      result = ERR_CREATELOG;
+      return result;
     }
+
     file.open(QIODevice::WriteOnly);
     QTextStream out(&file);
-    out << "#####" << '\n' << "Saved IC history from " << ao_app->current_server_name << " (" << ao_app->current_server_address << ") on " << QDateTime::currentDateTimeUtc().toString() << '\n' << "#####" << '\n';
-    foreach (chatlogpiece item, ic_chatlog_history) {
-        out << item.get_full() << '\n';
-      }
-
-    file.close();
-    if (file_exists(file.fileName()))
-      ic_result = SUCCESS;
-    else
-      ic_result = ERR_UNKNOWN;
-    return logfile;
-}
-
-QString Courtroom::save_ooc_chatlog()
-{
-    QString dtstamp_clean = QDateTime::currentDateTimeUtc().toString().replace(" ","-").replace(":",".");
-    QString logfile = "OOCLog-" + dtstamp_clean +  ".log";
-    ooc_result = ERR_UNKNOWN;
-    QDir logfolder("logs");
-    if (!logfolder.exists())
+    out << "#####" << '\n' << "Saved " + log_type + " history from " << ao_app->current_server_name << " (" << ao_app->current_server_address << ") on " << QDateTime::currentDateTimeUtc().toString() << '\n' << "#####" << '\n';
+    if (!is_ooc) 
     {
-      qDebug() << "I: directory ./logs does not exist, attempting to create";
-      QDir::current().mkdir(logfolder.dirName());
-      if (!logfolder.exists()) { // check again to be sure the folder was created
-        qDebug() << "W: attempted to create dir ./logs unsuccessfully";
-        ooc_result = ERR_CREATEDIR;
-        return logfile;
+      foreach (chatlogpiece item, ic_chatlog_history) 
```

Comment:
```suggestion
      for (const chatlogpiece &item : ic_chatlog_history) 
```

Reply:
I'll be honest - the underlying logic here scares me and I don't understand it. Making just the suggested change results in build errors, so I presume there are other changes to be made?

Reply:
This changes a deprecated use of the `foreach` macro to a safe equivalent (getting a const ref to an element). What build errors are you getting?

Reply:
```
\AO2-Client\src\text_file_functions.cpp:771: error: passing 'const chatlogpiece' as 'this' argument discards qualifiers [-fpermissive]
         out << item.get_full() << '\n';
                              ^
```

Reply:
Oh, that is because `get_full` needs to be marked as a `const` function (add `const` add the end of the function signature).

Reply:
The compiler yelled at me until I slapped `const` on nearly every function in `chatlogpiece.cpp`, seems like it works but I've still got my fingers crossed.